File: GDD_Twitch_API/TwitchAPIService.cs
using System;
using System.Configuration;
using System.Net.Http;
using System.Net.Http.Headers;
using System.Text.Json;
using System.Threading.Tasks;
using System.Collections.Generic;
using GGDTwitchAPI.Models;

namespace GGDTwitchAPI
{

    public class TwitchApiService
    {
        private readonly string _clientId;
        private readonly string _clientSecret;
        private string _accessToken;

        public TwitchApiService(string clientId, string clientSecret)
        {
            _clientId = clientId;
            _clientSecret = clientSecret;
        }

        public async Task InitializeAsync()
        {
            using var client = new HttpClient();
            var response = await client.PostAsync($"https://id.twitch.tv/oauth2/token" +
                $"?client_id={_clientId}&client_secret={_clientSecret}&grant_type=client_credentials", null);

            response.EnsureSuccessStatusCode();
            var json = JsonDocument.Parse(await response.Content.ReadAsStringAsync());
            _accessToken = json.RootElement.GetProperty("access_token").GetString();
        }

        public async Task<List<StreamerInfoModel>> GetInitialTwitchData(IEnumerable<string> namesOrIds, bool useIds = false)
        {
            var param = useIds ? "id" : "login";
            var query = string.Join("&", namesOrIds.Select(name => $"{param}={Uri.EscapeDataString(name)}"));

            using var client = CreateAuthorizedClient();
            var response = await client.GetAsync($"https://api.twitch.tv/helix/users?{query}");
            response.EnsureSuccessStatusCode();

            var json = JsonDocument.Parse(await response.Content.ReadAsStringAsync());
            return json.RootElement.GetProperty("data").EnumerateArray().Select(user => new StreamerInfoModel
            {
                Name = user.GetProperty("display_name").GetString(),
                ID = user.GetProperty("id").GetString()
            }).ToList();
        }
        private async Task<HttpResponseMessage> SendWithRetry(Func<HttpClient, Task<HttpResponseMessage>> action)
        {
            using var client = CreateAuthorizedClient();
            var response = await action(client);

            if (response.StatusCode == System.Net.HttpStatusCode.Unauthorized)
            {
                await InitializeAsync(); // refresh token
                using var retryClient = CreateAuthorizedClient();
                response = await action(retryClient);
            }

            response.EnsureSuccessStatusCode();
            return response;
        }

        public async Task<Dictionary<string, bool>> GetLiveStatus(List<string> streamerIds)
        {
            var query = string.Join("&", streamerIds.Select(id => $"user_id={id}"));

            var response = await SendWithRetry(client => client.GetAsync($"https://api.twitch.tv/helix/streams?{query}"));

            var json = JsonDocument.Parse(await response.Content.ReadAsStringAsync());
            var liveIds = json.RootElement.GetProperty("data")
                .EnumerateArray()
                .Select(entry => entry.GetProperty("user_id").GetString())
                .ToHashSet();

            return streamerIds.ToDictionary(id => id, id => liveIds.Contains(id));
        }

        public async Task<List<StreamerInfoModel>> GetStreamersWithLiveStatus(IEnumerable<string> namesOrIds, bool useIds = false)
        {
            var userInfo = await GetInitialTwitchData(namesOrIds, useIds);
            if (userInfo == null || userInfo.Count == 0)
                return new List<StreamerInfoModel>();

            var ids = userInfo.Select(u => u.ID).ToList();
            var liveStatus = await GetLiveStatus(ids);

            foreach (var user in userInfo)
            {
                user.IsLive = liveStatus.TryGetValue(user.ID, out bool isLive) && isLive;
            }

            return userInfo;
        }


        private HttpClient CreateAuthorizedClient()
        {
            if (string.IsNullOrEmpty(_accessToken))
            {
                throw new InvalidOperationException("Twitch API access token not initialized. Call InitializeAsync() first.");
            }

            var client = new HttpClient();
            client.DefaultRequestHeaders.Add("Authorization", $"Bearer {_accessToken}");
            client.DefaultRequestHeaders.Add("Client-Id", _clientId);
            return client;
        }

    }

}

File: GGD_LightController/TwitchAPI/twitch_api_service.py
import requests

class TwitchAPIService:
    def __init__(self, client_id, client_secret):
        self.client_id = client_id
        self.client_secret = client_secret
        self.access_token = None

    def initialize(self):
        url = "https://id.twitch.tv/oauth2/token"
        params = {
            "client_id": self.client_id,
            "client_secret": self.client_secret,
            "grant_type": "client_credentials"
        }
        response = requests.post(url, params=params)
        response.raise_for_status()
        self.access_token = response.json()["access_token"]

    def _get_headers(self):
        if not self.access_token:
            raise RuntimeError("Twitch API not initialized")
        return {
            "Client-ID": self.client_id,
            "Authorization": f"Bearer {self.access_token}"
        }

    def get_live_status(self, public_ids):
        def fetch_chunk(chunk):
            query = "&".join(f"user_id={id_}" for id_ in chunk)
            url = f"https://api.twitch.tv/helix/streams?{query}"
            response = requests.get(url, headers=self._get_headers())

            if response.status_code == 401:
                print("Access token expired — refreshing...")
                self.initialize()
                response = requests.get(url, headers=self._get_headers())

            response.raise_for_status()
            return response.json()

        live_status = {id_: False for id_ in public_ids}

        for chunk in [public_ids[i:i+100] for i in range(0, len(public_ids), 100)]:
            data = fetch_chunk(chunk)
            for stream in data.get("data", []):
                uid = stream["user_id"]
                live_status[uid] = True

        return live_status


